### PR TITLE
Better calculation of Delegation data

### LINKED ISSUE #263: Fix staking rewards query
This came from https://github.com/CosmWasm/cosmwasm/issues/513#issuecomment-682436132 and also requires https://github.com/CosmWasm/cosmwasm/issues/515 and https://github.com/CosmWasm/go-cosmwasm/issues/134 to properly implement the return value.

---

There is a [`FullDelegation` response](https://github.com/CosmWasm/cosmwasm/blob/c9a51417b5f9752fe466315bb198b80dc79dff3b/packages/std/src/query.rs#L146-L160), which does have an `accumulated_rewards` field, which is what you are looking for. 

If you look into the [go query handler](https://github.com/CosmWasm/wasmd/blob/master/x/wasm/internal/keeper/query_plugins.go#L164-L183), which calls into the Keeper to get the rewards. This is then [transformed into the cosmwasm type](https://github.com/CosmWasm/wasmd/blob/master/x/wasm/internal/keeper/query_plugins.go#L226-L234) to pass it into Rust. And *should* have the reward info.

Digging in, I see this was [left as a TODO](https://github.com/CosmWasm/wasmd/blob/master/x/wasm/internal/keeper/query_plugins.go#L230-L233) as I had no idea how to get such info from the staking module.

The implementation to calculate that was not done and needs to be implemented. Looking into this again, the calculation is done in [`distribution.Keeper.calculateDelegationRewards`](https://github.com/cosmos/cosmos-sdk/blob/v0.39.1/x/distribution/keeper/delegation.go#L56), which is a private method. Even the [query method](https://github.com/cosmos/cosmos-sdk/blob/v0.39.1/x/distribution/keeper/querier.go#L125-L157) is private. And many of the functions used inside are private.

### PR DESCRIPTION
**Pull request by ethanfrey**
Closes #263 

Hoping for https://github.com/cosmos/cosmos-sdk/issues/7466 to make `query_plugins.go :: getAccumulatedRewards` easier to write/maintain. But we cannot block on launchpad release, just be ready to update with a patch release when it is out.

### CODE REVIEW

**Comment by ethanfrey** [comment]
@alpe I think I finished this. Happy for a review. Using the distribution module is rather gnarly

File: x/wasm/internal/keeper/staking_test.go [hunk 1]
```diff
@@ -534,9 +534,71 @@ func TestQueryStakingInfo(t *testing.T) {
...
+	// note this is not bob (who staked to the contract), but the contract itself
+	require.Equal(t, contractAddr.String(), delInfo.Delegator)
+	// this is a different Coin type, with String not BigInt, compare field by field
+	require.Equal(t, funds[0].Denom, delInfo.Amount.Denom)
+	require.Equal(t, funds[0].Amount.String(), delInfo.Amount.Amount)
+
+	require.Equal(t, wasmTypes.NewCoin(200000, "stake"), delInfo.CanRedelegate)
+	require.Len(t, delInfo.AccumulatedRewards, 1)
+	// see bonding above to see how we calculate 36000 (240000 / 6 - 10% commission)
+	require.Equal(t, wasmTypes.NewCoin(36000, "stake"), delInfo.AccumulatedRewards[0])
```
Comment (alpe) [inline x/wasm/internal/keeper/staking_test.go:601]: pp: A test with receiving redelegations would be wonderful

**Review by alpe — APPROVED** [review]
Good work in a not so easy space.
There is only one question about the read only store that can be a blocker if it is writable store.

File: x/wasm/internal/keeper/query_plugins.go [hunk 2]
```diff
@@ -212,28 +209,82 @@ func sdkToDelegations(ctx sdk.Context, keeper staking.Keeper, delegations []stak
...
+	params := distribution.QueryDelegationRewardsParams{
+		DelegatorAddress: delegation.DelegatorAddress,
+		ValidatorAddress: delegation.ValidatorAddress,
+	}
+	data, err := json.Marshal(params)
+	if err != nil {
+		return nil, err
+	}
+	req := abci.RequestQuery{Data: data}
+	qres, err := distribution.NewQuerier(distKeeper)(ctx, []string{distribution.QueryDelegationRewards}, req)
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/wasm/internal/keeper/query_plugins.go:264]: The context is a CachedMultiStore, but if inside a Handler, this cache is written when the handler succeeds. 

Good point. I can do one more cachewrap around it just for the query and then unwrap at the end, ensuring the query data is not written.

File: x/wasm/internal/keeper/query_plugins.go [hunk 3]
```diff
@@ -212,28 +209,82 @@ func sdkToDelegations(ctx sdk.Context, keeper staking.Keeper, delegations []stak
...
 	amount := sdk.NewCoin(bondDenom, val.TokensFromShares(delegation.Shares).TruncateInt())
 
-	// can relegate? other query for redelegations?
-	// keeper.GetRedelegation
+	delegationCoins := convertSdkCoinToWasmCoin(amount)
+
+	// FIXME: this is very rough but better than nothing...
+	// if this (val, delegate) pair is receiving a redelegation, it cannot redelegate more
+	// otherwise, it can redelegate the full amount
+	// (there are cases of partial funds redelegated, but this is a start)
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/wasm/internal/keeper/query_plugins.go:225]: I mean, it is a fixme. `CanRedelegate` is <= actual amount that can be redelegated (better than always 0).

I have no idea how to fix this inside the staking/distribution handlers. That code is convoluted.

[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/wasm/internal/keeper/query_plugins.go:225]: Happy if you want to take a stab at improving this algorithm

File: x/wasm/internal/keeper/staking_test.go (hunk 1 above)
Comment (ethanfrey) [reply x/wasm/internal/keeper/staking_test.go:601]: Good point, but would you like to add one?

It took quite some time to get the setup working and I copied that from another test.
Happy for another PR that improves these tests

**Comment by ethanfrey** [comment]
@alpe there are two big follow-up tasks that do need doing. I made https://github.com/CosmWasm/wasmd/issues/282 to capture that. I can improve that issue, and if you want to aim for that for 0.11 you can. Otherwise, I will look more later.

File: x/wasm/internal/keeper/query_plugins.go [hunk 4]
```diff
@@ -212,28 +209,82 @@ func sdkToDelegations(ctx sdk.Context, keeper staking.Keeper, delegations []stak
...
+	redelegateCoins := wasmTypes.NewCoin(0, bondDenom)
+	if !keeper.HasReceivingRedelegation(ctx, delegation.DelegatorAddress, delegation.ValidatorAddress) {
```
[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/wasm/internal/keeper/query_plugins.go:227]: We already use it below in `Amount:    delegationCoins`. We need to calculate it in any case, we can use the same calculation for 2 return values possibly.

File: x/wasm/internal/keeper/query_plugins.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ethanfrey) [reply x/wasm/internal/keeper/query_plugins.go]: Fixed in https://github.com/CosmWasm/wasmd/pull/278/commits/aabe0b6287200749d3e1e7dca2ffab6f3c61324c

**Pull request merged**